clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2900,23 +2906,29 @@ class SYCLKernelNameTypeVisitor
       auto *NameSpace = dyn_cast_or_null<NamespaceDecl>(DeclCtx);
       if (NameSpace && NameSpace->isStdNamespace()) {
         S.Diag(KernelInvocationFuncLoc, diag::err_sycl_kernel_incorrectly_named)
-            << /* kernel name cannot be a type in the std namespace */ 3;
+            << QualType(Tag->getTypeForDecl(), 0)
+            << /* kernel name cannot be a type in the std namespace */ 2;
+        S.Diag(KernelInvocationFuncLoc, diag::note_kernel_name)
+            << KernelNameType;
         IsInvalid = true;
         return;
       }
       if (!DeclCtx->isTranslationUnit() && !isa<NamespaceDecl>(DeclCtx)) {
         const bool KernelNameIsMissing = Tag->getName().empty();
         if (KernelNameIsMissing) {
-          S.Diag(KernelInvocationFuncLoc,
-                 diag::err_sycl_kernel_incorrectly_named)
-              << /* kernel name is missing */ 0;
+          S.Diag(KernelInvocationFuncLoc, diag::err_sycl_kernel_name_missing);
```

Comment:
Question to reviewers: Do we need to emit the problematic type in the error diagnostic for this case as well?
I will be updating the diagnostic message in an upcoming patch from "Kernel name is missing" to "Unnamed type used in a SYCL kernel name" and would like to know if this note will be sufficient.

Reply:
I'm not sure that note like `expected-note-re@Inputs/sycl.hpp:220 {{Invalid kernel name is '(lambda at {{.*}}unnamed-kernel.cpp{{.*}}'}}` (copy paste from the test) looks useful. `Unnamed type used in a SYCL kernel name` is also looks better than `Kernel name is missing`

Reply:
Do you prefer to remove the note and instead add the invalid type in the error diagnostic like other cases?

Reply:
@srividya-sundaram FYI I'm changing the diagnostics as per this comment here - https://github.com/intel/llvm/pull/2730#discussion_r517866371